File: cpp/simpleon.cpp
#include "simpleon.hpp"
#include <vector>
#include <iostream>
#include <deque>
#include <iomanip>

using namespace simpleon;
using namespace std;

string IData_DUMMY_STRING;
list<IData *> IData_DUMMY_LIST;
map<string, IData *> IData_DUMMY_MAP;

bool IData::GetBool() { return false; }
int IData::GetInt() { return 0; }
double IData::GetFloat() { return 0.0; }
const string & IData::GetString() { return IData_DUMMY_STRING; }
const list<IData *> & IData::GetList() { return IData_DUMMY_LIST; }
const map<string, IData *> & IData::GetDict() { return IData_DUMMY_MAP; }

class NullData : public IData {
public:
    Type GetType() override { return T_NULL; }
};

class IntData : public IData {
public:
    int value;
    Type GetType() override { return T_INT; }
    int GetInt() override { return value; } 
};

class BoolData : public IData {
public:
    bool value;
    Type GetType() override { return T_BOOL; }
    bool GetBool() override { return value; } 
};

class FloatData : public IData {
public:
    double value;
    Type GetType() override { return T_FLOAT; }
    double GetFloat() override { return value; } 
};

class StringData : public IData {
private:
    bool _quoted;
public:
    StringData(bool quoted) : _quoted(quoted) {}
    string value;
    Type GetType() override { return _quoted ? T_STRING : T_UQ_STRING; }
    const string & GetString() override { return value; } 
};

class ListData : public IData {
public:
    list<IData *> value;
    Type GetType() override { return T_LIST; }
    const list<IData *> & GetList() override { return value; }
    ~ListData() override {
        for (auto d : value)
            delete d;
    }
};

class DictData : public IData {
public:
    map<string, IData *> value;
    Type GetType() override { return T_DICT; }
    const map<string, IData *> & GetDict() override { return value; }
    ~DictData() override {
        for (auto && kv : value)
            delete kv.second;
    }
};

ParseException::ParseException(const char * w) : _what(w) { }
const char * ParseException::what() const noexcept { return _what.c_str(); }

template<typename I>
struct CharMap {
    I m[256];

    CharMap(const char * v, I defaultValue, I setValue) {
        for (int i = 0; i < 256; ++i) m[i] = defaultValue;
        while (*v) {
            m[*v] = setValue;
            ++v;
        }
    }

    CharMap(const char * v, I defaultValue, const initializer_list<I> & l) {
        for (int i = 0; i < 256; ++i) m[i] = defaultValue;
        auto it = begin(l);
        while (*v && it != end(l)) {
            m[*v] = *it;
            ++v; ++it;
        }
    }
};

static const CharMap<bool> SPECIAL_CHARS(" \t[]{}:\",#", false, true);
static const CharMap<bool> NUM_CHARS("+-.0123456789", false, true);
static const CharMap<int> HEX_TRANSLATE("0123456789abcdefABCDEF", -1,
    {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 10, 11, 12, 13, 14, 15}
);

#define IS_SPECIAL_CHAR(c) (SPECIAL_CHARS.m[c])
#define BUF_CLEAN_THRESHOLD 4096

class SimpleONParser : public IParser {
private:

    enum State {
        STATE_ELEMENT_START,
        STATE_ELEMENT_END,
        STATE_DICT_PRE_KEY,
        STATE_DICT_KEY,
        STATE_DICT_POST_KEY,
        STATE_DICT_VALUE,
        STATE_DICT_POST_VALUE,
        STATE_LIST,
        STATE_QUOTED_STRING,
        STATE_QUOTELESS_STRING,
        STATE_MULTILINE_STRING
    };
    
    bool            _convert;
    bool            _multi;
    bool            _sealed;
    string          _buf;
    size_t          _readPos;
    vector<string>  _keyStack;
    vector<IData *> _valueStack;
    vector<State>   _stateStack;
    deque<IData *>  _results;
    
public:

    SimpleONParser(bool convert, bool multi) {
        _convert = convert;
        _multi = multi;
        _sealed = false;
        _readPos = 0;
        _stateStack.push_back(STATE_ELEMENT_START);
    }
    
    void ParseLine(const string & line) override {
        if (_sealed) return;

        _buf.append(line);
        ParseBuf();
    }

    void Seal() override {
        if (_sealed) return;

        _sealed = true;
        ParseBuf();

        _readPos = 0;
        _buf.clear();
        _stateStack.clear();
        for (auto d : _valueStack) {
            delete d;
        }
        _valueStack.clear();
    }

    ~SimpleONParser() {
        Seal();
        for (auto d : _results) {
            delete d;
        }
    }

    void HandleEscape() {
        if (_readPos >= _buf.size()) return;
        auto && value = (*(StringData *)_valueStack.back()).value;
        switch (_buf[_readPos]) {
        case 'n':
            value += '\n';
            ++_readPos;
            break;
        case 't':
            value += '\t';
            ++_readPos;
            break;
        case 'r':
            value += '\r';
            ++_readPos;
            break;
        case 'f':
            value += '\f';
            ++_readPos;
            break;
        case 'x':
        {
            unsigned char c;
            int digit;

            if (_readPos + 2 >= _buf.size()) goto InputError;
            ++_readPos;
            digit = HEX_TRANSLATE.m[_buf[_readPos]]; if (digit < 0) goto InputError;
            c = digit; ++_readPos;
            digit = HEX_TRANSLATE.m[_buf[_readPos]]; if (digit < 0) goto InputError;
            c = (c << 4) + digit; ++_readPos;
            value += c;

            break;

        InputError:
            throw ParseException("Expect 2 hex chars for utf-8 escape");
        }
        case '/':
        case '\\':
        case '"':
            value += _buf[_readPos];
            ++_readPos;
            break;
        default:
            value += '\\';
            break;
        }
    }

    void HandleUnquotedString(size_t s, size_t e, bool convert) {
        if (convert) {
            string word(_buf, s, e - s);
            if (NUM_CHARS.m[_buf[s]]) {
                int vint;
                bool succ = true;
                size_t idx;

                try {
                    vint = stoi(word, &idx);
                    if (idx != e - s) succ = false;
                }
                catch (const exception &) {
                    succ = false;
                }

                if (succ) {
                    auto v = new IntData();
                    v->value = vint;
                    _valueStack.push_back(v);
                    return;
                }

                double dint;
                succ = true;

                try {
                    dint = stod(word, &idx);
                    if (idx != e - s) succ = false;
                }
                catch (const exception &) {
                    succ = false;
                }

                if (succ) {
                    auto v = new FloatData();
                    v->value = dint;
                    _valueStack.push_back(v);
                    return;
                }
            }
            else if (word[0] == 'n' && word == "null") {
                auto v = new NullData();
                _valueStack.push_back(v);
                return;
            }
            else if (word[0] == 't' && word == "true") {
                auto v = new BoolData();
                v->value = true;
                _valueStack.push_back(v);
                return;
            }
            else if (word[0] == 'f' && word == "false") {
                auto v = new BoolData();
                v->value = false;
                _valueStack.push_back(v);
                return;
            }
        }

        auto v = new StringData(false);
        v->value.assign(_buf, s, e - s);
        _valueStack.push_back(v);
    }

    void CleanBuf() {
        if (_readPos > _buf.size()) _readPos = _buf.size(); 
        if (_readPos > BUF_CLEAN_THRESHOLD) {
            _buf = _buf.substr(_readPos, _buf.size() - _readPos);
            _readPos = 0;
        }
    }

    void ParseBuf() {
        size_t limit = _buf.size();
        while (_readPos < limit) {
            if (_stateStack.size() == 0) {
                Seal();
                return;
            }

            auto state = _stateStack.back();
            
            switch (state) {
            case STATE_ELEMENT_END: {
                auto value = _valueStack.back();
                
                _stateStack.pop_back();
                _valueStack.pop_back();
				
				if (_stateStack.size() == 0) {
                    _results.push_back(value);

                    if (_multi) {
                        _stateStack.push_back(STATE_ELEMENT_START);
                    }

                    break;
                }
                
                switch (_stateStack.back()) {
                case STATE_DICT_KEY:
                    _stateStack.back() = STATE_DICT_POST_KEY;
                    _keyStack.push_back(value->GetString());
                    delete value;
                    break;
                    
                case STATE_DICT_VALUE:
                    ((DictData *)_valueStack.back())->value[_keyStack.back()] = value;
                    _keyStack.pop_back();
                    _stateStack.back() = STATE_DICT_POST_VALUE;
                    break;
                    
                case STATE_LIST:
                    ((ListData *)_valueStack.back())->value.push_back(value);
                    break;

                default:
                    throw ParseException("invalid state to insert element");
                }
                break;
            }
            case STATE_QUOTED_STRING: {
                size_t s = _readPos;
                while (s < limit && _buf[s] != '"' && _buf[s] != '\\') ++s;

                ((StringData *)_valueStack.back())->value.append(_buf, _readPos, s - _readPos);
                
                if (s >= limit) {
                    _readPos = limit;
                    _stateStack.back() = STATE_ELEMENT_END;
                }
                else {
                    if (_buf[s] == '\\') {
                        _readPos = s + 1;
                        HandleEscape();
                    }
                    else {
                        _readPos = s + 1;
                        _stateStack.back() = STATE_ELEMENT_END;
                    }
                }
                break;
            }
            case STATE_MULTILINE_STRING: {
                size_t s = _readPos;
                while (s < limit && _buf[s] != '"' && _buf[s] != '\\') ++s;
                ((StringData *)_valueStack.back())->value.append(_buf, _readPos, s - _readPos);
                
                if (s >= limit) {
                    ((StringData *)_valueStack.back())->value += '\n';
                    _readPos = limit;
                }
                else {
                    if (_buf[s] == '\\') {
                        _readPos = s + 1;
                        HandleEscape();
                    }
                    else if (s + 2 < limit && _buf[s + 1] == '"' && _buf[s + 2] == '"') {
                        _readPos = s + 3;
                        _stateStack.back() = STATE_ELEMENT_END;
                    }
                    else {
                        ((StringData *)_valueStack.back())->value += '"';
                        _readPos = s + 1;
                    }
                }
                break;
            }
            case STATE_DICT_PRE_KEY: {
                size_t s = _readPos;
                while (s < limit && (_buf[s] == ' ' || _buf[s] == '\t')) ++s;

                if (s >= limit) {
                    _readPos = limit;
                }
                else if (_buf[s] == '"' || !IS_SPECIAL_CHAR(_buf[s])) {
                    _readPos = s;
                    _stateStack.back() = STATE_DICT_KEY;
                    _stateStack.push_back(STATE_ELEMENT_START);
                }
                else if (_buf[s] == '}') {
                    _stateStack.back() = STATE_ELEMENT_END;
                    _readPos = s + 1;
                }
                else if (_buf[s] == '#') {
                    _readPos = limit;
                }
                else {
                    throw ParseException("format error - expecting dict key or end");
                }
                
                break;
            }
            case STATE_DICT_KEY: {
                break;
            }
            case STATE_DICT_POST_KEY: {
                size_t s = _readPos;
                while (s < limit && (_buf[s] == ' ' || _buf[s] == '\t')) ++s;

                if (s >= limit) {
                    _readPos = limit;
                }
                else if (_buf[s] == ':') {
                    _readPos = s + 1;
                    _stateStack.back() = STATE_DICT_VALUE;
                    _stateStack.push_back(STATE_ELEMENT_START);
                }
                else if (_buf[s] == '#') {
                    _readPos = limit;
                }
                else {
                    throw ParseException("format error - expecting key-value-separator");
                }
                
                break;
            }
            case STATE_DICT_VALUE: {
                break;
            }
            case STATE_DICT_POST_VALUE: {
                size_t s = _readPos;
                while (s < limit && (_buf[s] == ' ' || _buf[s] == '\t')) ++s;

                if (s >= limit) {
                    _readPos = limit;
                }
                else if (_buf[s] == '"' || !IS_SPECIAL_CHAR(_buf[s])) {
                    _stateStack.back() = STATE_DICT_PRE_KEY;
                    _readPos = s;
                }
                else if (_buf[s] == ',') {
                    _stateStack.back() = STATE_DICT_PRE_KEY;
                    _readPos = s + 1;
                }
                else if (_buf[s] == '}') {
                    _stateStack.back() = STATE_ELEMENT_END;
                    _readPos = s + 1;
                }
                else if (_buf[s] == '#') {
                    _readPos = limit;
                }
                else {
                    throw ParseException("format error in dict");
                }
                
                break;
            }
            case STATE_LIST: {
                size_t s = _readPos;
                while (s < limit && (_buf[s] == ' ' || _buf[s] == '\t')) ++s;

                if (s >= limit) {
                    _readPos = limit;
                }
                else if (_buf[s] == ']') {
                    _stateStack.back() = STATE_ELEMENT_END;
                    _readPos = s + 1;
                }
                else if (_buf[s] == ',') {
                    _readPos = s + 1;
                    _stateStack.push_back(STATE_ELEMENT_START);
                }
                else if (_buf[s] == '#') {
                    _readPos = limit;
                }
                else {
                    _readPos = s;
                    _stateStack.push_back(STATE_ELEMENT_START);
                }
                
                break;
            }
            case STATE_ELEMENT_START: {
                size_t s = _readPos;
                while (s < limit && (_buf[s] == ' ' || _buf[s] == '\t')) ++s;

                if (s >= limit) {
                    _readPos = limit;
                }
                else if (_buf[s] == '{') {
                    _valueStack.push_back(new DictData());
                    _stateStack.back() = STATE_DICT_PRE_KEY;
                    _readPos = s + 1;
                }
                else if (_buf[s] == '[') {
                    _valueStack.push_back(new ListData());
                    _stateStack.back() = STATE_LIST;
                    _readPos = s + 1;
                }
                else if (_buf[s] == '"') {
                    _valueStack.push_back(new StringData(true));
                    if (s + 2 < limit &&
                        _buf[s + 1] == '"' && _buf[s + 2] == '"') {
                        _stateStack.back() = STATE_MULTILINE_STRING;
                        _readPos = s + 3;
                    }
                    else {
                        _stateStack.back() = STATE_QUOTED_STRING;
                        _readPos = s + 1;
                    }
                }
                else if (_buf[s] == '#') {
                    _readPos = limit;
                }
                else {
                    size_t e = s;
                    while (e < limit && !IS_SPECIAL_CHAR(_buf[e])) ++e;

                    if (_stateStack.size() > 1 && _stateStack[_stateStack.size() - 2] == STATE_DICT_KEY) {
                        HandleUnquotedString(s, e, false);
                    }
                    else {
                        HandleUnquotedString(s, e, _convert);
                    }

                    _stateStack.back() = STATE_ELEMENT_END;
                    _readPos = e;
                }
                
                break;
            }
            }
        }

        CleanBuf();
    }
    
    IData * Extract() override {
        if (_results.size() > 0) {
            auto v = _results.front();
            _results.pop_front();
            return v;
        }
        else {
            return nullptr;
        }
    }
};

IParser * simpleon::CreateSimpleONParser(bool convert, bool multi) {
    return new SimpleONParser(convert, multi);
}

void simpleon::Dump(ostream & o, IData * d) {
    if (d == nullptr) {
        o << "(Null)";
        return;
    }

    switch (d->GetType()) {
    case IData::T_NULL:
        o << "null";
        break;
    case IData::T_BOOL:
        o << d->GetBool() ? "true" : "false";
        break;
    case IData::T_INT:
        o << d->GetInt();
        break;
    case IData::T_FLOAT:
        o << d->GetFloat();
        break;
    case IData::T_STRING:
    case IData::T_UQ_STRING:
        o << quoted(d->GetString());
        break;
    case IData::T_LIST: {
        o << '[';
        bool first = true;
        for (IData * ele : d->GetList()) {
            if (first) first = false;
            else o << ',';
            Dump(o, ele);
        }
        o << ']';
        break;
    }
    case IData::T_DICT: {
        o << '{';
        bool first = true;
        for (auto && kv : d->GetDict()) {
            if (first) first = false;
            else o << ',';
            o << quoted(kv.first) << ':';
            Dump(o, kv.second);
        }
        o << '}';
        break;
    }
    }
}

File: python/simpleon-py/simpleon.py
import re
import struct
from collections import deque

def default_token_handler(s):
    if s[0] in "-0123456789.":
        try:
            return int(s)
        except ValueError:
            pass

        try:
            return float(s)
        except ValueError:
            pass

    if s[0] == "t" and s == "true":
        return True

    if s[0] == "f" and s == "false":
        return False
    
    if s[0] == "n" and s == "null":
        return None

    return s
    
class ParseException(Exception):
    pass

class SimpleONParser:

    STATE_ELEMENT_START    = 0
    STATE_ELEMENT_END      = 1
    STATE_DICT_PRE_KEY     = 2
    STATE_DICT_KEY         = 3
    STATE_DICT_POST_KEY    = 4
    STATE_DICT_VALUE       = 5
    STATE_DICT_POST_VALUE  = 6
    STATE_LIST             = 7
    STATE_QUOTED_STRING    = 8
    STATE_QUOTELESS_STRING = 9
    STATE_MULTILINE_STRING = 10

    # FOR DEBUGGING:
    # STATE_ELEMENT_START    = "ES"
    # STATE_ELEMENT_END      = "EE"
    # STATE_DICT_PRE_KEY     = "K-"
    # STATE_DICT_KEY         = "K:"
    # STATE_DICT_POST_KEY    = "K+"
    # STATE_DICT_VALUE       = "V:"
    # STATE_DICT_POST_VALUE  = "V+"
    # STATE_LIST             = "L"
    # STATE_QUOTED_STRING    = "QS"
    # STATE_QUOTELESS_STRING = "QL"
    # STATE_MULTILINE_STRING = "ML"

    COMMENT_CHAR_RE = re.compile(r'#')
    QUOTED_STRING_SPECIAL_RE = re.compile(r'[\\"]')
    MULTILINE_STRING_SPECIAL_RE = re.compile(r'\\|"""')
    UNQUOTED_RE = re.compile(r'[^\[\]{}:", \t]+')
    DICT_KEY_VALUE_SEP_RE = re.compile(":")
    ITEM_SEP_RE = re.compile(",")
    NON_WHITESPACE_RE = re.compile(r'[^ \t]')

    BUF_CLEAN_SIZE = 4096
    
    def __init__(self, multi = True, to_convert = True, token_handler = default_token_handler):
        self.reset(multi, to_convert, token_handler)
        
    def reset(self, multi = True, to_convert = True, token_handler = default_token_handler):
        self.sealed = False
        self.multi = multi
        self.to_convert = to_convert
        self.state_stack = [ self.STATE_ELEMENT_START ];
        self.value_stack = [ None ];
        self.buf = ""
        self.buf_read_pos = 0
        self.token_handler = token_handler
        self.values = deque()
    
    def seal(self):
        if self.sealed:
            return
        self.sealed = True
        self.state_stack = [ ];
        self.value_stack = [ ];
        self.buf = ""
    
    def state_get(self):
        if len(self.state_stack) == 0:
            raise ParseException("stack is empty")
        return self.state_stack[len(self.state_stack) - 1]

    def state_set(self, state):
        if len(self.state_stack) == 0:
            raise ParseException("stack is empty")
        self.state_stack[len(self.state_stack) - 1] = state

    def value_get(self):
        if len(self.value_stack) == 0:
            raise ParseException("stack is empty")
        return self.value_stack[len(self.value_stack) - 1]

    def value_set(self, value):
        if len(self.value_stack) == 0:
            raise ParseException("stack is empty")
        self.value_stack[len(self.value_stack) - 1] = value

    def pop(self):
        if len(self.state_stack) == 0:
            raise ParseException("stack is empty")
        self.state_stack.pop()
        self.value_stack.pop()

    def push(self, state, value):
        self.state_stack.append(state)
        self.value_stack.append(value)

    def parse(self, s):
        self.parse_lines(s.splitlines())
        return self.extract()

    def parse_lines(self, lines):
        for line in lines:
            self.parse_line(line)
        pass

    def parse_line(self, line):
        if self.sealed:
            return
        self.buf += line
        self.parse_buf()
        
    def handle_escape(self):
        start_c = self.buf[self.buf_read_pos]
        if start_c == 'n':
            self.value_get().extend(b'\n')
            self.buf_read_pos += 1
        elif start_c == 't':
            self.value_get().extend(b'\t')
            self.buf_read_pos += 1
        elif start_c == 'b':
            self.value_get().extend(b'\b')
            self.buf_read_pos += 1
        elif start_c == 'f':
            self.value_get().extend(b'\f')
            self.buf_read_pos += 1
        elif start_c == 'x':
            if self.buf_read_pos + 2 >= len(self.buf):
                raise ParseException("expect 2 hex chars for utf-8 escaping")
            code_str = self.buf[self.buf_read_pos + 1:self.buf_read_pos + 3]
            try:
                code = int(code_str, 16)
            except Exception:
                raise ParseException("expect 2 hex chars for utf-8 escaping (parse {0} failed)".format(code_str))
            self.value_get().append(code)
            self.buf_read_pos += 3
        elif start_c == '"' or start_c == '\\' or start_c == '/':
            self.value_get().append(ord(start_c))
            self.buf_read_pos += 1
        else:
            self.value_get().extend(b"\\")
            self.value_get().append(ord(start_c))
            
    def clean_buf(self):
        if self.buf_read_pos > len(self.buf):
            self.buf_read_pos = len(self.buf)
        if self.buf_read_pos > self.BUF_CLEAN_SIZE:
            self.buf = self.buf[self.buf_read_pos:]
            self.buf_read_pos = 0
        
    def parse_buf(self):
        state = None
        while len(self.buf) > self.buf_read_pos or state == self.STATE_ELEMENT_END:
        
            self.clean_buf()
            
            if len(self.state_stack) == 0:
                self.seal()
                return

            state = self.state_get()
            current = self.value_get()
            read_pos = self.buf_read_pos
            # print((state, current, read_pos))

            if state == self.STATE_ELEMENT_END:
                value = current
                if isinstance(value, bytearray):
                    value = value.decode("utf-8")
                self.pop()
                
                if len(self.state_stack) == 0:
                    self.values.append(value)
                    
                    if self.multi:
                        self.push(self.STATE_ELEMENT_START, None)
                    
                    continue

                state = self.state_get()
                current = self.value_get()

                if state == self.STATE_DICT_KEY:
                    state = self.STATE_DICT_POST_KEY
                    current = [current, value]
                elif state == self.STATE_DICT_VALUE:
                    current[0][current[1]] = value
                    state = self.STATE_DICT_POST_VALUE
                    current = current[0]
                elif state == self.STATE_LIST:
                    current.append(value)
                else:
                    raise ParseException("invalid position to insert element")
                
            elif state == self.STATE_QUOTED_STRING:
                m = self.QUOTED_STRING_SPECIAL_RE.search(self.buf, read_pos)
                if not m:
                    current.extend(self.buf[read_pos:].encode("utf-8"))
                    state = self.STATE_ELEMENT_END
                    read_pos = len(self.buf)
                else:
                    current.extend(self.buf[read_pos:m.start(0)].encode("utf-8"))
                    if self.buf[m.start(0)] == "\\":
                        self.buf_read_pos = m.start(0) + 1
                        self.handle_escape()
                        continue
                    elif self.buf[m.start(0)] == '"':
                        # the string is ended
                        read_pos = m.start(0) + 1 
                        state = self.STATE_ELEMENT_END
                    else:
                        raise ParseException("format error in quoted string")

            elif state == self.STATE_MULTILINE_STRING:
                m = self.MULTILINE_STRING_SPECIAL_RE.search(self.buf, read_pos)
                if not m:
                    current.extend(self.buf[read_pos:].encode("utf-8"))
                    current.extend(b"\n")
                    read_pos = len(self.buf)
                else:
                    current.extend(self.buf[read_pos:m.start(0)].encode("utf-8"))
                    if self.buf[m.start(0)] == "\\":
                        self.buf_read_pos = m.start(0) + 1
                        self.handle_escape()
                        continue
                    elif self.buf[m.start(0):m.start(0) + 3] == '"""':
                        # the string is ended
                        read_pos = m.start(0) + 3
                        state = self.STATE_ELEMENT_END
                    else:
                        raise ParseException("format error in multi-line string")

            elif state == self.STATE_DICT_PRE_KEY:
                m = self.NON_WHITESPACE_RE.search(self.buf, read_pos)
                if not m:
                    read_pos = len(self.buf)
                elif self.buf[m.start(0)] == '"' or self.UNQUOTED_RE.match(self.buf[m.start(0)]):
                    self.buf_read_pos = m.start(0)
                    self.state_set(self.STATE_DICT_KEY)
                    self.push(self.STATE_ELEMENT_START, None)
                    continue
                elif self.buf[m.start(0)] == "}":
                    state = self.STATE_ELEMENT_END
                    read_pos = m.start(0) + 1
                elif self.buf[m.start(0)] == "#":
                    read_pos = len(self.buf)
                else:
                    raise ParseException("format error - expecting dict key or end")

            elif state == self.STATE_DICT_KEY:
                pass

            elif state == self.STATE_DICT_POST_KEY:
                m = self.NON_WHITESPACE_RE.search(self.buf, read_pos)
                if not m:
                    read_pos = len(self.buf)
                elif self.DICT_KEY_VALUE_SEP_RE.match(self.buf[m.start(0)]):
                    self.buf_read_pos = m.start(0) + 1
                    self.state_set(self.STATE_DICT_VALUE)
                    self.push(self.STATE_ELEMENT_START, None)
                    continue
                elif self.buf[m.start(0)] == "#":
                    read_pos = len(self.buf)                
                else:
                    raise ParseException("format error - expecting key-value-separator")
            
            elif state == self.STATE_DICT_VALUE:
                pass

            elif state == self.STATE_DICT_POST_VALUE:
                m = self.NON_WHITESPACE_RE.search(self.buf, read_pos)
                if not m:
                    read_pos = len(self.buf)
                elif self.buf[m.start(0)] == '"' or self.UNQUOTED_RE.match(self.buf[m.start(0)]):
                    state = self.STATE_DICT_PRE_KEY
                    read_pos = m.start(0)
                elif self.ITEM_SEP_RE.match(self.buf[m.start(0)]):
                    state = self.STATE_DICT_PRE_KEY
                    read_pos = m.start(0) + 1
                elif self.buf[m.start(0)] == "}":
                    state = self.STATE_ELEMENT_END
                    read_pos = m.start(0) + 1
                elif self.buf[m.start(0)] == "#":
                    read_pos = len(self.buf)
                else:
                    raise ParseException("format error in dict")

            elif state == self.STATE_LIST:
                m = self.NON_WHITESPACE_RE.search(self.buf, read_pos)
                if not m:
                    read_pos = len(self.buf)
                elif self.buf[m.start(0)] == "]":
                    state = self.STATE_ELEMENT_END
                    read_pos = m.start(0) + 1
                elif self.ITEM_SEP_RE.match(self.buf[m.start(0)]):
                    self.buf_read_pos = m.start(0) + 1
                    self.push(self.STATE_ELEMENT_START, None)
                    continue
                elif self.buf[m.start(0)] == "#":
                    read_pos = len(self.buf)
                else:
                    self.buf_read_pos = m.start(0)
                    self.push(self.STATE_ELEMENT_START, None)
                    continue

            elif state == self.STATE_ELEMENT_START:
                m = self.NON_WHITESPACE_RE.search(self.buf, read_pos)
                if not m:
                    read_pos = len(self.buf)
                elif self.buf[m.start(0)] == "{":
                    current = dict()
                    state = self.STATE_DICT_PRE_KEY
                    read_pos = m.start(0) + 1
                elif self.buf[m.start(0)] == "[":
                    current = list()
                    state = self.STATE_LIST
                    read_pos = m.start(0) + 1
                elif self.buf[m.start(0)] == '"':
                    current = bytearray()
                    if self.buf[m.start(0):m.start(0) + 3] == '"""':
                        state = self.STATE_MULTILINE_STRING
                        read_pos = m.start(0) + 3
                    else:
                        state = self.STATE_QUOTED_STRING
                        read_pos = m.start(0) + 1
                elif self.buf[m.start(0)] == "#":
                    read_pos = len(self.buf)
                elif self.UNQUOTED_RE.match(self.buf[m.start(0)]):
                    m_end = self.UNQUOTED_RE.search(self.buf, m.start(0))
                    if not m_end:
                        current = self.buf[m.start(0):]
                        state = self.STATE_ELEMENT_END
                        read_pos = len(self.buf)
                    else:
                        current = self.buf[m_end.start(0):m_end.end(0)]
                        state = self.STATE_ELEMENT_END
                        read_pos = m_end.end(0)

                    if len(current) == 0:
                        raise ParseException("unexpect char at element start")
                    
                    if self.to_convert and \
                        (    len(self.state_stack) <= 1 \
                        or self.state_stack[len(self.state_stack) - 2] != self.STATE_DICT_KEY):
                        current = self.token_handler(current)
                else:
                    raise ParseException("format error in parsing general element")

            self.state_set(state)
            self.value_set(current)
            self.buf_read_pos = read_pos
            
        self.clean_buf()
        
    def extract(self):
        if len(self.values) == 0:
            return None
        else:
            return self.values.popleft()
